scripts/testing/database-emulator-test.js

```diff
@@ -0,0 +1,140 @@
+      const writeStream = fs.createWriteStream(filepath);
+      console.log(`Downloading emulator from [${EMULATOR_LINK}] ...`);
+      request(EMULATOR_LINK)
+        .pipe(writeStream)
+        .on('finish', () => {
+          resolve(filepath);
+          console.log(`Saved emulator binary file to [${filepath}].`);
+        })
+        .on('error', reject);
+    });
+  });
+}
+
+async function launchEmulator(filepath) {
+  return new Promise((resolve, reject) => {
+    const promise = spawn(
+      'java',
+      ['-jar', path.basename(filepath), '--port', EMULATOR_PORT],
+      {
+        cwd: path.dirname(filepath),
+        stdio: 'inherit'
+      }
+    );
+    promise.catch(reject);
+
+    const timeout = 10; // seconds
+    console.log(`Waiting for emulator to start up ...`);
+    const start = Date.now();
+    setTimeout(function wait() {
+      if (Date.now() - start > timeout * 1000) {
+        reject(`Emulator not ready after ${timeout}s. Exiting ...`);
+      } else {
+        console.log(`Ping emulator at [http://localhost:${EMULATOR_PORT}] ...`);
+        request(`http://localhost:${EMULATOR_PORT}`, (error, response) => {
+          if (error && error.code === 'ECONNREFUSED') {
+            setTimeout(wait, 1000);
+          } else if (response && response.statusCode === 400) {
+            console.log('Emulator has started up successfully!');
+            resolve(promise.childProcess);
+          } else {
+            reject(error);
+          }
+        });
+      }
+    }, 1000);
+  });
+}
+
+async function loadPublicRules() {
+  console.log('Loading rule {".read": true, ".write": true} to emulator ...');
+  return new Promise((resolve, reject) => {
+    request.put(
+      {
+        uri: `http://localhost:${EMULATOR_PORT}/.settings/rules.json?ns=${EMULATOR_NAMESPACE}`,
+        headers: { Authorization: 'Bearer owner' },
+        body: '{ "rules": { ".read": true, ".write": true } }'
+      },
+      (error, response, body) => {
+        if (error) reject(error);
+        console.log(body);
+        resolve(response.statusCode);
+      }
+    );
+  });
+}
+
+async function runTest() {
+  const options = {
+    cwd: path.resolve(__dirname, '../../packages/database'),
+    env: Object.assign({}, process.env, {
+      USE_RTDB_EMULATOR: true,
+      RTDB_EMULATOR_PORT: EMULATOR_PORT,
+      RTDB_EMULATOR_NAMESPACE: EMULATOR_NAMESPACE
+    }),
+    stdio: 'inherit'
+  };
+  return spawn('yarn', ['test'], options);
+}
+
+async function start() {
+  let emulator;
+  try {
+    const filepath = await downloadEmulator();
+    emulator = await launchEmulator(filepath);
+    await loadPublicRules();
+    await runTest();
+  } catch (err) {
+    console.error(err);
+    process.exitCode = 1;
+  } finally {
+    if (emulator) {
+      console.log(`Shutting down emulator, pid: [${emulator.pid}] ...`);
+      emulator.kill();
+    }
+  }
+}
+
+start().catch(err => {
+  console.error(err);
```

Comment:
This catch block never runs unless `emulator.kill()` throws. Is it your intention?

Reply:
Yes. I'm also wondering if I should omit the catch block in function 'start', and handle all the errors here. I'm not sure which one is better.

Reply:
I think I would remove the catch { } block above and move the `process.exitCode = 1;` to here.

Reply:
Done.